lib/tracecontext/tracestate.js

```diff
@@ -0,0 +1,309 @@
+/**
+ * Class for Managing Tracestate
+ *
+ * Class that creates objects for managing trace state.
+ * This class is capable of parsing both tracestate strings
+ * and tracestate binary representations, allowing clients
+ * to get and set values in a single list-member/namespace
+ * while preserving values in the other namespaces.
+ *
+ * Capable of working with either the binary of string
+ * formatted tracestate values.
+ *
+ * Usage:
+ *   const tracestate = TraceState.fromStringFormatString(headerTracestate, 'es')
+ *   tracestate.setValue('s',1)
+ *   const newHeader = tracestate.toW3cString()
+ */
+class TraceState {
+  constructor (sourceBuffer, listMemberNamespace = 'es', defaultValues = {}, logger) {
+    if (!this._validateVendorKey(listMemberNamespace)) {
+      throw new Error('Vendor namespace failed validation.')
+    }
+
+    // buffer representation of the other vendor namespace
+    this.buffer = sourceBuffer
+
+    // values for our namespace
+    this.values = defaultValues
+
+    this.listMemberNamespace = listMemberNamespace
+
+    // if no logger, then fallback to agent singleton
+    logger = logger || require('../..').logger
+    this.logger = logger
+  }
+
+  setValue (key, value) {
+    const strKey = String(key)
+    const strValue = String(value)
+    if (!this._validateElasicKeyAndValue(strKey, strValue)) {
+      this.logger.trace('could not set tracestate key, invaliad characters detected')
+      return false
+    }
+
+    const oldValue = this.values[strKey]
+    this.values[strKey] = value
+
+    // if new length is greater than 256, undo the setting
+    const serializedValue = this._serializeValues(this.values)
```

Comment:
I am kind of confused with this interpretation. https://www.w3.org/TR/trace-context/#tracestate-limits

Do we need to account for this during setting or during serialisation?

Reply:
_nod_ -- this isn't straight forward, but [the elastic distributed tracing spec](https://github.com/elastic/apm/blob/d5b2c87326548befcfec6731713932a00e430b99/specs/agents/tracing-distributed-tracing.md) includes the following.

> If adding another key/value pair to the es entry would exceed the limit of 256 chars, that key/value pair MUST be ignored by agents. The key/value and entry separators : and ; have to be considered as well.

If the serialized value is greater than 256 characters, then the value needs to be ignored.  We _could_ handle this purely at serialization time by using a `Map` to store keys (to ensure their order), and then serialize until things got too long.  However, doing so would mean `setValue` would return `true` for a value that ends up not being set (which @jahtalab pointed out this code fails to do, which I intend to fix)   

That's why I chose this implementation -- it's _slightly_ more work each time a value is set, but we're able to signal out that a value reached the limit.  

My preference would be to keep this implementation, but it's only a slight preference.  If you think we need something else let's talk offline/slack about it.